File: native-wrapper/ios-native/GnhSecurity/GnhBiometricModule.swift
import Foundation
import LocalAuthentication
import Security

/** Native-only biometric gate — encrypts secrets in Keychain; no export to JS. */
final class GnhBiometricModule {
  static let purposeApp = "app"
  static let purposeData = "data"

  private let prefs = GnhSecurePrefs()
  private var inFlight = false
  private let keyService = "im.getnowhere.app.gnh.biometric"

  func isAvailable(purpose: String) -> Bool {
    guard purpose == Self.purposeApp || purpose == Self.purposeData else { return false }
    let ctx = LAContext()
    var error: NSError?
    return ctx.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error)
  }

  func enrollDataUnlock(walletId: String, password: String, completion: @escaping ([String: Any]) -> Void) {
    guard beginPrompt() else { completion(["error": "busy"]); return }
    guard isAvailable(purpose: Self.purposeData) else { endPrompt(); completion(["error": "unsupported"]); return }
    let credentialId = randomCredentialId(prefix: "data-")
    storeProtected(credentialId: credentialId, plaintext: password) { [weak self] ok in
      guard let self else { return }
      if ok {
        self.prefs.set(key: self.metaKey(credentialId), value: walletId)
        self.endPrompt()
        completion(["credentialId": credentialId])
      } else {
        self.removeCredential(credentialId: credentialId)
        self.endPrompt()
        completion(["error": "cancelled"])
      }
    }
  }

  func unlockDataUnlock(walletId: String, credentialId: String, completion: @escaping ([String: Any]) -> Void) {
    guard beginPrompt() else { completion(["error": "busy"]); return }
    guard prefs.get(key: metaKey(credentialId)) == walletId else {
      endPrompt(); completion(["error": "failed"]); return
    }
    loadProtected(credentialId: credentialId) { [weak self] plaintext in
      self?.endPrompt()
      if let plaintext { completion(["password": plaintext]) }
      else { completion(["error": "cancelled"]) }
    }
  }

  func enrollAppAccess(passcode: String, completion: @escaping ([String: Any]) -> Void) {
    guard beginPrompt() else { completion(["error": "busy"]); return }
    guard isAvailable(purpose: Self.purposeApp) else { endPrompt(); completion(["error": "unsupported"]); return }
    let credentialId = randomCredentialId(prefix: "app-")
    storeProtected(credentialId: credentialId, plaintext: passcode) { [weak self] ok in
      guard let self else { return }
      if ok {
        self.prefs.set(key: "gnh.appAccessCredentialId", value: credentialId)
        self.endPrompt()
        completion(["credentialId": credentialId])
      } else {
        self.removeCredential(credentialId: credentialId)
        self.endPrompt()
        completion(["error": "cancelled"])
      }
    }
  }

  func unlockAppAccess(completion: @escaping ([String: Any]) -> Void) {
    guard beginPrompt() else { completion(["error": "busy"]); return }
    guard let credentialId = prefs.get(key: "gnh.appAccessCredentialId") else {
      endPrompt(); completion(["error": "failed"]); return
    }
    loadProtected(credentialId: credentialId) { [weak self] _ in
      self?.endPrompt()
      completion(["ok": true])
    }
  }

  func removeCredential(credentialId: String) {
    deleteKeychainItem(account: credentialId)
    prefs.remove(key: metaKey(credentialId))
    if prefs.get(key: "gnh.appAccessCredentialId") == credentialId {
      prefs.remove(key: "gnh.appAccessCredentialId")
    }
  }

  private func beginPrompt() -> Bool {
    if inFlight { return false }
    inFlight = true
    return true
  }

  private func endPrompt() { inFlight = false }

  private func storeProtected(credentialId: String, plaintext: String, completion: @escaping (Bool) -> Void) {
    let ctx = LAContext()
    ctx.localizedReason = "Confirm biometrics"
    var error: NSError?
    guard ctx.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) else {
      completion(false); return
    }
    ctx.evaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, localizedReason: "Enable biometric unlock") { ok, _ in
      guard ok else { completion(false); return }
      let data = Data(plaintext.utf8)
      self.deleteKeychainItem(account: credentialId)
      var query = self.keychainQuery(account: credentialId)
      query[kSecValueData as String] = data
      let status = SecItemAdd(query as CFDictionary, nil)
      completion(status == errSecSuccess)
    }
  }

  private func loadProtected(credentialId: String, completion: @escaping (String?) -> Void) {
    let ctx = LAContext()
    var query = keychainQuery(account: credentialId)
    query[kSecReturnData as String] = true
    query[kSecMatchLimit as String] = kSecMatchLimitOne
    query[kSecUseAuthenticationContext as String] = ctx
    ctx.localizedReason = "Unlock with biometrics"
    var item: CFTypeRef?
    let status = SecItemCopyMatching(query as CFDictionary, &item)
    if status == errSecSuccess, let data = item as? Data, let text = String(data: data, encoding: .utf8) {
      completion(text)
    } else {
      completion(nil)
    }
  }

  private func keychainQuery(account: String) -> [String: Any] {
    var access: SecAccessControl?
    var error: Unmanaged<CFError>?
    access = SecAccessControlCreateWithFlags(
      nil,
      kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly,
      .biometryCurrentSet,
      &error,
    )
    var query: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: keyService,
      kSecAttrAccount as String: account,
    ]
    if let access { query[kSecAttrAccessControl as String] = access }
    return query
  }

  private func deleteKeychainItem(account: String) {
    let query: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: keyService,
      kSecAttrAccount as String: account,
    ]
    SecItemDelete(query as CFDictionary)
  }

  private func metaKey(_ credentialId: String) -> String { "meta_\(credentialId)" }

  private func randomCredentialId(prefix: String) -> String {
    prefix + UUID().uuidString.replacingOccurrences(of: "-", with: "").prefix(16)
  }
}

File: native-wrapper/android-native/GnhSecurity/GnhBiometricModule.kt
package im.getnowhere.app.security

import android.content.Context
import android.os.Build
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyPermanentlyInvalidatedException
import android.security.keystore.KeyProperties
import android.util.Base64
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import org.json.JSONObject
import java.security.KeyStore
import java.security.SecureRandom
import java.util.concurrent.Executor
import java.util.concurrent.atomic.AtomicBoolean
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

/**
 * Native-only biometric gate: encrypts wallet password or app passcode in Keystore.
 * No intermediate secrets cross the WebView bridge.
 */
class GnhBiometricModule(
    private val activity: FragmentActivity,
    private val securePrefs: GnhSecurePrefs,
) {
    private val inFlight = AtomicBoolean(false)

    fun isAvailable(purpose: String): Boolean {
        if (purpose != PURPOSE_APP && purpose != PURPOSE_DATA) return false
        return isBiometricAvailable()
    }

    fun enrollDataUnlock(walletId: String, password: String, callback: (JSONObject) -> Unit) {
        if (!beginPrompt()) {
            callback(error("busy"))
            return
        }
        if (!isBiometricAvailable()) {
            endPrompt()
            callback(error("unsupported"))
            return
        }
        val credentialId = randomCredentialId(PREFIX_DATA)
        runCryptoEnroll(
            credentialId = credentialId,
            title = "Enable biometric unlock",
            subtitle = "Confirm with fingerprint or face",
            plaintext = password.toByteArray(Charsets.UTF_8),
            onSuccess = {
                securePrefs.set(metaKey(credentialId), walletId)
                endPrompt()
                callback(ok().put("credentialId", credentialId))
            },
            onError = { code ->
                removeStoredCredential(credentialId)
                securePrefs.remove(metaKey(credentialId))
                endPrompt()
                callback(error(code))
            },
        )
    }

    fun unlockDataUnlock(walletId: String, credentialId: String, callback: (JSONObject) -> Unit) {
        if (!beginPrompt()) {
            callback(error("busy"))
            return
        }
        if (securePrefs.get(metaKey(credentialId)) != walletId) {
            endPrompt()
            callback(error("failed"))
            return
        }
        runCryptoUnlock(
            credentialId = credentialId,
            title = "Unlock wallet",
            subtitle = "Use biometrics to unlock",
            onSuccess = { plaintext ->
                endPrompt()
                callback(ok().put("password", String(plaintext, Charsets.UTF_8)))
            },
            onError = { code ->
                if (code == "invalidated") removeStoredCredential(credentialId)
                endPrompt()
                callback(error(code))
            },
        )
    }

    fun enrollAppAccess(passcode: String, callback: (JSONObject) -> Unit) {
        if (!beginPrompt()) {
            callback(error("busy"))
            return
        }
        if (!isBiometricAvailable()) {
            endPrompt()
            callback(error("unsupported"))
            return
        }
        val credentialId = randomCredentialId(PREFIX_APP)
        runCryptoEnroll(
            credentialId = credentialId,
            title = "Enable app unlock",
            subtitle = "Confirm with fingerprint or face",
            plaintext = passcode.toByteArray(Charsets.UTF_8),
            onSuccess = {
                securePrefs.set(APP_CREDENTIAL_KEY, credentialId)
                endPrompt()
                callback(ok().put("credentialId", credentialId))
            },
            onError = { code ->
                removeStoredCredential(credentialId)
                securePrefs.remove(APP_CREDENTIAL_KEY)
                endPrompt()
                callback(error(code))
            },
        )
    }

    fun unlockAppAccess(callback: (JSONObject) -> Unit) {
        if (!beginPrompt()) {
            callback(error("busy"))
            return
        }
        val credentialId = securePrefs.get(APP_CREDENTIAL_KEY)
        if (credentialId.isNullOrEmpty()) {
            endPrompt()
            callback(error("failed"))
            return
        }
        runCryptoUnlock(
            credentialId = credentialId,
            title = "Unlock app",
            subtitle = "Use biometrics",
            onSuccess = {
                endPrompt()
                callback(ok().put("ok", true))
            },
            onError = { code ->
                if (code == "invalidated") {
                    removeStoredCredential(credentialId)
                    securePrefs.remove(APP_CREDENTIAL_KEY)
                }
                endPrompt()
                callback(error(code))
            },
        )
    }

    fun removeCredential(credentialId: String) {
        removeStoredCredential(credentialId)
        securePrefs.remove(metaKey(credentialId))
        if (securePrefs.get(APP_CREDENTIAL_KEY) == credentialId) {
            securePrefs.remove(APP_CREDENTIAL_KEY)
        }
    }

    private fun beginPrompt(): Boolean = inFlight.compareAndSet(false, true)

    private fun endPrompt() {
        inFlight.set(false)
    }

    private fun isBiometricAvailable(): Boolean {
        val manager = BiometricManager.from(activity)
        var authenticators = BiometricManager.Authenticators.BIOMETRIC_STRONG
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            authenticators = authenticators or BiometricManager.Authenticators.DEVICE_CREDENTIAL
        }
        return manager.canAuthenticate(authenticators) == BiometricManager.BIOMETRIC_SUCCESS
    }

    private fun runCryptoEnroll(
        credentialId: String,
        title: String,
        subtitle: String,
        plaintext: ByteArray,
        onSuccess: () -> Unit,
        onError: (String) -> Unit,
    ) {
        activity.runOnUiThread {
            try {
                val key = createKey(credentialId)
                val cipher = Cipher.getInstance("AES/GCM/NoPadding")
                cipher.init(Cipher.ENCRYPT_MODE, key)
                showCryptoPrompt(
                    title,
                    subtitle,
                    cipher,
                    onAuthenticated = {
                        try {
                            val iv = cipher.iv
                            val ciphertext = cipher.doFinal(plaintext)
                            saveBlob(credentialId, iv, ciphertext)
                            onSuccess()
                        } catch (_: Exception) {
                            onError("failed")
                        }
                    },
                    onError = onError,
                )
            } catch (_: Exception) {
                onError("failed")
            }
        }
    }

    private fun runCryptoUnlock(
        credentialId: String,
        title: String,
        subtitle: String,
        onSuccess: (ByteArray) -> Unit,
        onError: (String) -> Unit,
    ) {
        activity.runOnUiThread {
            try {
                val blob = loadBlob(credentialId) ?: run {
                    onError("failed")
                    return@runOnUiThread
                }
                val cipher = Cipher.getInstance("AES/GCM/NoPadding")
                val spec = GCMParameterSpec(GCM_TAG_LENGTH, blob.iv)
                val key = loadKey(credentialId)
                cipher.init(Cipher.DECRYPT_MODE, key, spec)
                showCryptoPrompt(
                    title,
                    subtitle,
                    cipher,
                    onAuthenticated = {
                        try {
                            onSuccess(cipher.doFinal(blob.ciphertext))
                        } catch (e: Exception) {
                            if (e.cause is KeyPermanentlyInvalidatedException) onError("invalidated")
                            else onError("failed")
                        }
                    },
                    onError = onError,
                )
            } catch (e: Exception) {
                if (e is KeyPermanentlyInvalidatedException) onError("invalidated")
                else onError("failed")
            }
        }
    }

    private fun showCryptoPrompt(
        title: String,
        subtitle: String,
        cipher: Cipher,
        onAuthenticated: () -> Unit,
        onError: (String) -> Unit,
    ) {
        val executor: Executor = ContextCompat.getMainExecutor(activity)
        val prompt = BiometricPrompt(
            activity,
            executor,
            object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                    onAuthenticated()
                }

                override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                    onError("cancelled")
                }
            },
        )
        val infoBuilder = BiometricPrompt.PromptInfo.Builder()
            .setTitle(title)
            .setSubtitle(subtitle)
            .setNegativeButtonText("Cancel")
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            infoBuilder.setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG)
        }
        try {
            prompt.authenticate(infoBuilder.build(), BiometricPrompt.CryptoObject(cipher))
        } catch (_: Exception) {
            onError("failed")
        }
    }

    private fun createKey(credentialId: String): SecretKey {
        val alias = keystoreAlias(credentialId)
        val keyStore = KeyStore.getInstance(KEYSTORE).apply { load(null) }
        if (keyStore.containsAlias(alias)) keyStore.deleteEntry(alias)
        val generator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, KEYSTORE)
        val builder = KeyGenParameterSpec.Builder(
            alias,
            KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT,
        )
            .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
            .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
            .setUserAuthenticationRequired(true)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            builder.setUserAuthenticationParameters(0, KeyProperties.AUTH_BIOMETRIC_STRONG)
        } else {
            builder.setUserAuthenticationValidityDurationSeconds(-1)
        }
        generator.init(builder.build())
        return generator.generateKey()
    }

    private fun loadKey(credentialId: String): SecretKey {
        val keyStore = KeyStore.getInstance(KEYSTORE).apply { load(null) }
        val entry = keyStore.getEntry(keystoreAlias(credentialId), null) as KeyStore.SecretKeyEntry
        return entry.secretKey
    }

    private fun saveBlob(credentialId: String, iv: ByteArray, ciphertext: ByteArray) {
        securePrefs.set(prefKey(credentialId, "iv"), b64(iv))
        securePrefs.set(prefKey(credentialId, "ct"), b64(ciphertext))
    }

    private fun loadBlob(credentialId: String): EncryptedBlob? {
        val iv = securePrefs.get(prefKey(credentialId, "iv")) ?: return null
        val ct = securePrefs.get(prefKey(credentialId, "ct")) ?: return null
        return EncryptedBlob(b64d(iv), b64d(ct))
    }

    private fun removeStoredCredential(credentialId: String) {
        try {
            val keyStore = KeyStore.getInstance(KEYSTORE).apply { load(null) }
            val alias = keystoreAlias(credentialId)
            if (keyStore.containsAlias(alias)) keyStore.deleteEntry(alias)
        } catch (_: Exception) {
        }
        securePrefs.remove(prefKey(credentialId, "iv"))
        securePrefs.remove(prefKey(credentialId, "ct"))
    }

    private data class EncryptedBlob(val iv: ByteArray, val ciphertext: ByteArray)

    companion object {
        private const val KEYSTORE = "AndroidKeyStore"
        private const val GCM_TAG_LENGTH = 128
        private const val PREFIX_DATA = "data-"
        private const val PREFIX_APP = "app-"
        const val PURPOSE_APP = "app"
        const val PURPOSE_DATA = "data"
        private const val APP_CREDENTIAL_KEY = "gnh.appAccessCredentialId"

        fun create(context: Context, activity: FragmentActivity): GnhBiometricModule {
            return GnhBiometricModule(activity, GnhSecurePrefs(context))
        }

        private fun randomCredentialId(prefix: String): String =
            prefix + b64(randomBytes(12)).replace("=", "")

        private fun randomBytes(n: Int): ByteArray =
            ByteArray(n).also { SecureRandom().nextBytes(it) }

        private fun keystoreAlias(credentialId: String) =
            "gnh_bio_" + credentialId.replace(Regex("[^A-Za-z0-9_-]"), "_")

        private fun prefKey(credentialId: String, suffix: String) = "cred_${credentialId}_$suffix"

        private fun metaKey(credentialId: String) = "meta_$credentialId"

        private fun b64(bytes: ByteArray) =
            Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)

        private fun b64d(value: String) =
            Base64.decode(value, Base64.URL_SAFE or Base64.NO_PADDING)

        private fun ok(): JSONObject = JSONObject()

        private fun error(code: String): JSONObject = JSONObject().put("error", code)
    }
}
